awx/playbooks/action_plugins/insights.py

```diff
@@ -34,24 +35,65 @@ def write_version(self, proj_path, etag):
         with open(file_path, 'w') as f:
             f.write(etag)
 
+    def _obtain_auth_token(self, oidc_endpoint, client_id, client_secret, result):
+        main_url = urljoin(oidc_endpoint, '/.well-known/openid-configuration')
+        response = requests.get(url=main_url, headers={'Accept': 'application/json'})
+        if response.status_code != 200:
+            result['failed'] = True
+            result['msg'] = 'Expected {} to return a status code of 200 but returned status code "{}" instead with content "{}".'.format(
+                main_url, response.status_code, response.content
+            )
+            return result
+
+        auth_url = response.json().get('token_endpoint', None)
+        data = {
+            'grant_type': 'client_credentials',
+            'scope': 'api.console',
+            'client_id': client_id,
+            'client_secret': client_secret,
+        }
+        response = requests.post(url=auth_url, data=data)
+
+        if response.status_code != 200:
+            result['failed'] = True
+            result['msg'] = 'Expected {} to return a status code of 200 but returned status code "{}" instead with content "{}".'.format(
+                auth_url, response.status_code, response.content
+            )
+        else:
+            result['token'] = response.json().get('access_token', None)
+            result['token_type'] = response.json().get('token_type', None)
+        return result
+
     def run(self, tmp=None, task_vars=None):
         self._supports_check_mode = False
 
         result = super(ActionModule, self).run(tmp, task_vars)
 
         insights_url = self._task.args.get('insights_url', None)
-        username = self._task.args.get('username', None)
-        password = self._task.args.get('password', None)
         proj_path = self._task.args.get('project_path', None)
         license = self._task.args.get('awx_license_type', None)
         awx_version = self._task.args.get('awx_version', None)
+        authentication = self._task.args.get('authentication', None)
+        username = self._task.args.get('username', None)
+        password = self._task.args.get('password', None)
+        client_id = self._task.args.get('client_id', None)
+        client_secret = self._task.args.get('client_secret', None)
+        oidc_endpoint = self._task.args.get('oidc_endpoint', None)
 
-        session = requests.Session()
-        session.auth = requests.auth.HTTPBasicAuth(username, password)
         headers = {
             'Content-Type': 'application/json',
             'User-Agent': '{} {} ({})'.format('AWX' if license == 'open' else 'Red Hat Ansible Automation Platform', awx_version, license),
         }
+
+        session = requests.Session()
+        if authentication == 'service_account' or (client_id and client_secret):
+            self._obtain_auth_token(oidc_endpoint, client_id, client_secret, result)
```

Comment:
I don't love passing the dict `result` into `_obtain_auth_token()`. Instead, can `_obtain_auth_token` return a tuple `(token_type, token)` or `None`.

Right now, with `result` passed in and being returned from the module, I'm thinking the `result['token']` might can end up in the Ansible output long with `msg`. We wouldn't want that behavior.